source/eventcore/drivers/posix/processes.d

```diff
@@ -10,427 +10,441 @@ import std.algorithm.comparison : among;
+		return ret;
+	}
+
+	final override void addRef(ProcessID pid)
+	{
+		lockedProcessInfo(pid, (info) {
+			nogc_assert(info.refCount > 0, "Adding reference to unreferenced process FD.");
+			info.refCount++;
+		});
+	}
+
+	final override bool releaseRef(ProcessID pid)
+	{
+		bool ret;
+		lockedProcessInfo(pid, (info) {
+			nogc_assert(info.refCount > 0, "Releasing reference to unreferenced process FD.");
+			if (--info.refCount == 0) {
+				// Remove/deallocate process
+				if (info.userDataDestructor)
+					() @trusted { info.userDataDestructor(info.userData.ptr); } ();
+
+				() @trusted { s_processes.remove(pid); } ();
+				ret = false;
+			} else ret = true;
+		});
+		return ret;
+	}
+
+	final protected override void* rawUserData(ProcessID pid, size_t size, DataInitializer initialize, DataInitializer destroy)
+	@system {
+		void* ret;
+		lockedProcessInfo(pid, (info) @safe nothrow {
+			assert(info.userDataDestructor is null || info.userDataDestructor is destroy,
+				"Requesting user data with differing type (destructor).");
+			assert(size <= ProcessInfo.userData.length, "Requested user data is too large.");
+
+			if (!info.userDataDestructor) {
+				() @trusted { initialize(info.userData.ptr); } ();
+				info.userDataDestructor = destroy;
+			}
+			ret = () @trusted { return info.userData.ptr; } ();
+		});
+		return ret;
+	}
+
+	package final @property size_t pendingCount() const nothrow @trusted { return s_processes.length; }
+
+
+	shared static this()
+	{
+		s_mutex = new shared Mutex;
+	}
+
+	private static void lockedProcessInfo(ProcessID pid, scope void delegate(ProcessInfo*) nothrow @safe fn)
+	{
+		s_mutex.lock_nothrow();
+		scope (exit) s_mutex.unlock_nothrow();
+		auto info = () @trusted { return pid in s_processes; } ();
+		fn(info);
+	}
+
+		private static void add(ProcessID pid, ProcessInfo info) @trusted {
+			s_mutex.lock_nothrow();
+			scope (exit) s_mutex.unlock_nothrow();
+
+			if (!s_waitThread) {
+				s_waitThread = new Thread(&waitForProcesses);
+				s_waitThread.start();
+			}
+
+			assert(pid !in s_processes, "Process adopted twice");
+			s_processes[pid] = info;
+		}
+
+		private static void waitForProcesses()
+		@system {
+			import core.stdc.errno : ECHILD, errno;
+			import core.sys.posix.sys.wait : idtype_t, WNOHANG, WNOWAIT, WEXITED, WEXITSTATUS, WIFEXITED, WTERMSIG, waitid, waitpid;
+			import core.sys.posix.signal : siginfo_t;
+
+			while (true) {
+				siginfo_t dummy;
+				auto ret = waitid(idtype_t.P_ALL, -1, &dummy, WEXITED|WNOWAIT);
+				if (ret == -1) {
+					{
+						s_mutex.lock_nothrow();
+						scope (exit) s_mutex.unlock_nothrow();
+						s_waitThread = null;
+					}
+					break;
+				}
+
+				ProcessID[] allprocs;
+
+				{
+					s_mutex.lock_nothrow();
+					scope (exit) s_mutex.unlock_nothrow();
+
+
+					() @trusted {
+						foreach (ref entry; s_processes.byKeyValue) {
+							if (!entry.value.exited)
+								allprocs ~= entry.key;
+						}
+					} ();
+				}
+
+				foreach (pid; allprocs) {
```

Comment:
> I might be wrong here, but couldn't a process go out of reference at this point, causing lockedProcessInfo to get a null ProcessInfo*, resulting in a segfault? We should at least add an assert into lockedProcessInfo to make sure the pointer is not null.

There is a `if (info is null)` check at line 371 (`onProcessExitStatic`), which *should* catch that case, if I'm not overlooking something.

> I think this also begs the question of what we should do if a process is not waited on, ie. the last reference is lost before it exits. Maybe it's worth putting an assert in releaseRef to make sure the last reference is lost after the process has completed so that zombies are easier to debug.

I didn't notice it before starting to work on this PR, but the reference handling in general goes against the usual rules where `releaseRef` must be the final call to free up the associated slot. The changes required to fix this are large enough that I'd like to split this into a separate PR, though, also considering that this issue already exists in the current master version.

BTW, I think the reason why the sequence `spawn` -> `wait` -> `releaseRef` currently works is that the initial ref count is zero and wraps around to `size_t.max` after the `wait` is done, so that finally the `releaseRef` call decrements it to `size_t.,max - 1` without asserting. It means that currently all slots will leak and finally crash once a PID gets reused.